src/Domain/CoordinateMaps/CylindricalEndcap.cpp

```diff
@@ -0,0 +1,914 @@
+    : center_(center),
+      radius_(radius),
+      theta_(acos((x_plane - center_[0]) / radius_)) {}
+
+double InnerSphereMap::sin_ax_over_x(double x, double ax, double a) noexcept {
+  return square(ax) < 6.0 * std::numeric_limits<double>::epsilon()
+             ? a
+             : sin(ax) / x;
+}
+DataVector InnerSphereMap::sin_ax_over_x(const DataVector& x,
+                                         double a) noexcept {
+  DataVector result(x);
+  for (size_t i = 0; i < x.size(); ++i) {
+    result[i] = sin_ax_over_x(x[i], a * x[i], a);
+  }
+  return result;
+}
+
+double InnerSphereMap::dlogx_sin_ax_over_x(double x, double ax,
+                                           double a) noexcept {
+  return square(ax) < 10.0 * std::numeric_limits<double>::epsilon()
+             ? -cube(a) / 3.0
+             : (a * cos(ax) - sin(ax) / x) / square(x);
+}
+DataVector InnerSphereMap::dlogx_sin_ax_over_x(const DataVector& x,
+                                               double a) noexcept {
+  DataVector result(x);
+  for (size_t i = 0; i < x.size(); ++i) {
+    result[i] = dlogx_sin_ax_over_x(x[i], a * x[i], a);
+  }
+  return result;
+}
+
+template <typename T>
+std::array<tt::remove_cvref_wrap_t<T>, 3> InnerSphereMap::operator()(
+    const std::array<T, 3>& source_coords) const noexcept {
+  using return_type = tt::remove_cvref_wrap_t<T>;
+  const return_type& xbar = source_coords[0];
+  const return_type& ybar = source_coords[1];
+  const return_type rho = sqrt(square(xbar) + square(ybar));
+  const return_type sin_factor = radius_ * sin_ax_over_x(rho, theta_);
+  const return_type x = radius_ * cos(rho * theta_) + center_[0];
+  const return_type y = sin_factor * xbar + center_[1];
+  const return_type z = sin_factor * ybar + center_[2];
+  return std::array<return_type, 3>{{std::move(x), std::move(y), std::move(z)}};
+}
+
+boost::optional<std::array<double, 3>> InnerSphereMap::inverse(
+    const std::array<double, 3>& target_coords) const noexcept {
+  const double x = target_coords[0] - center_[0];
+  const double y = target_coords[1] - center_[1];
+  const double z = target_coords[2] - center_[2];
+  const double r = sqrt(square(x) + square(y) + square(z));
+  if (not equal_within_roundoff(r, radius_)) {
+    return boost::none;
+  }
+
+  // Compute sin^2(rho theta).
+  const double sin_squared_rho_theta = (square(y) + square(z)) / square(r);
+  // Compute sin(rho theta)/rho.
+  // If sin^2(rho theta) is small,
+  // use arcsin(q) = q(1 + q^2/6 + 3 q^4/40 + ...)
+  // for q = sin(rho theta).
+  double sin_rho_theta_over_rho = 0.0;
+  if (square(sin_squared_rho_theta) <
+      (40.0 / 3.0) * std::numeric_limits<double>::epsilon()) {
+    sin_rho_theta_over_rho = theta_ * (1.0 - sin_squared_rho_theta / 6.0);
+  } else {
+    const double rho = asin(sqrt(sin_squared_rho_theta)) / theta_;
+    sin_rho_theta_over_rho = sqrt(sin_squared_rho_theta) / rho;
+  }
+
+  const double xbar = y / (r * sin_rho_theta_over_rho);
```

Comment:
Can `r` or `sin_rho_theta_over_rho` be zero at this point?

Reply:
No they cannot. I will add a comment as to why.